### PR TITLE
Deprecate SSHKeys in favor of SSHKeysRef variable

### PR DESCRIPTION
**Pull request by Prajyot-Parab**
Signed-off-by: Prajyot-Parab <prajyot.parab2@ibm.com>

**What this PR does / why we need it**:
- Deprecate SSHKeys in favor of SSHKeysRef variable

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

/area provider/ibmcloud

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note
Deprecate SSHKeys in favor of SSHKeysRef variable
```

### CODE REVIEW

File: api/v1beta1/ibmvpcmachine_types.go [hunk 1]
```diff
@@ -58,7 +58,28 @@ type IBMVPCMachineSpec struct {
 	PrimaryNetworkInterface NetworkInterface `json:"primaryNetworkInterface,omitempty"`
 
 	// SSHKeys is the SSH pub keys that will be used to access VM.
+	// Deprecated: Use SSHKeysRef instead
+	// +optional
 	SSHKeys []*string `json:"sshKeys,omitempty"`
+
+	// SSHKeysRef is the SSH pub keys that will be used to access VM.
+	// +optional
+	SSHKeysRef []*IBMVPCResourceReference `json:"sshKeysRef,omitempty"`
```
Comment (Amulyam24) [inline api/v1beta1/ibmvpcmachine_types.go:67]: One question here, why are both `SSHKeys` and `SSHKeysRef` marked as optional here? In a case where both are nil, it is currently not handled while creating the instance in `machine.go`-L150-174 below.

File: cloud/scope/machine.go [hunk 2]
```diff
@@ -155,6 +155,22 @@ func (m *MachineScope) CreateMachine() (*vpcv1.Instance, error) {
 			}
 			instancePrototype.Keys = append(instancePrototype.Keys, key)
 		}
+	} else if m.IBMVPCMachine.Spec.SSHKeysRef != nil {
+		instancePrototype.Keys = []vpcv1.KeyIdentityIntf{}
+		for i, sshKey := range m.IBMVPCMachine.Spec.SSHKeysRef {
+			keyID, err := fetchKeyID(sshKey, m)
+			if err != nil {
+				record.Warnf(m.IBMVPCMachine, "FailedRetriveKey", "Failed key retrival - %v", err)
```
Comment (Amulyam24) [inline cloud/scope/machine.go:163]: small nit(typo) - retrieve & retrieval

File: api/v1beta1/ibmvpcmachine_types.go (hunk 1 above)
Comment (Amulyam24) [reply api/v1beta1/ibmvpcmachine_types.go:67]: got the reason, this comment can be ignored.

File: cloud/scope/machine.go (hunk 2 above)
Comment (Prajyot-Parab) [reply cloud/scope/machine.go]: Fixed.

**Review by Amulyam24 — APPROVED** [review]
/lgtm

File: cloud/scope/machine.go [hunk 3]
```diff
@@ -155,6 +155,19 @@ func (m *MachineScope) CreateMachine() (*vpcv1.Instance, error) {
 			}
 			instancePrototype.Keys = append(instancePrototype.Keys, key)
 		}
+	} else if m.IBMVPCMachine.Spec.SSHKeysRef != nil {
```
Comment (mkumatag) [inline cloud/scope/machine.go:158]: but we discussed about using both values `SSHKeys SSHKeysRef` and right?

File: cloud/scope/machine.go [hunk 4]
```diff
@@ -155,6 +155,19 @@ func (m *MachineScope) CreateMachine() (*vpcv1.Instance, error) {
...
+		instancePrototype.Keys = []vpcv1.KeyIdentityIntf{}
+		for _, sshKey := range m.IBMVPCMachine.Spec.SSHKeysRef {
+			keyID, err := fetchKeyID(sshKey, m)
+			if err != nil {
+				record.Warnf(m.IBMVPCMachine, "FailedRetrieveKey", "Failed key retrieval - %v", err)
```
Comment (mkumatag) [inline cloud/scope/machine.go:163]: should be error here right? which key? can this be more specific to ssh key?

File: cloud/scope/machine.go [hunk 5]
```diff
@@ -327,3 +340,30 @@ func (m *MachineScope) GetBootstrapData() (string, error) {
...
+	if key.Name != nil {
+		keys, _, err := m.IBMVPCClient.ListKeys(&vpcv1.ListKeysOptions{})
+		if err != nil {
+			m.Logger.Error(err, "Failed to get keys")
+			return nil, err
+		}
+
+		for _, k := range keys.Keys {
+			if *k.Name == *key.Name {
+				m.Logger.Info("Key found with ID", "Key", *k.Name, "ID", *k.ID)
```
Comment (mkumatag) [inline cloud/scope/machine.go:358]: this seems like a debug log and shouldn't be enabled by default.

File: cloud/scope/machine.go [hunk 6]
```diff
@@ -327,3 +340,30 @@ func (m *MachineScope) GetBootstrapData() (string, error) {
...
+				return k.ID, nil
+			}
+		}
+	}
+
+	if key.ID != nil && key.Name != nil {
+		return nil, fmt.Errorf("both ID and Name can't be nil")
+	}
+
+	return nil, fmt.Errorf("failed to find an Key ID")
```
Comment (mkumatag) [inline cloud/scope/machine.go:368]: This error is not clear, why it is not able to find Key ID?

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: cloud/scope/machine.go (hunk 3 above)
Comment (Prajyot-Parab) [reply cloud/scope/machine.go]: Fixed.

File: cloud/scope/machine.go (hunk 5 above)
Comment (Prajyot-Parab) [reply cloud/scope/machine.go]: Fixed.

File: cloud/scope/machine.go (hunk 4 above)
Comment (Prajyot-Parab) [reply cloud/scope/machine.go]: Fixed.

File: cloud/scope/machine.go (hunk 6 above)
Comment (Prajyot-Parab) [reply cloud/scope/machine.go]: Fixed.

**Review by mkumatag — COMMENTED** [review]
overall lgtm, how about adding some UT around this?

**Comment by Prajyot-Parab** [comment]
> overall lgtm, how about adding some UT around this?

Done.

**Review by mkumatag — APPROVED** [review]
/lgtm

**Pull request merged**